### ISSUE TITLE
GPG key distribution support

### ISSUE THREAD

**Issue by githubatf2f10** [mail-in-a-box/mailinabox#1877]
Hi Josh,

I was wondering whether it's a small effort that you may have get the following features integrated.
1. OpenPGP DNS RR: https://weberblog.net/pgp-key-distribution-via-dnssec-openpgpkey/
2. GPG WKD: https://wiki.gnupg.org/WKDHosting

These are two methods that we may use to help facilitate our own GPG key distribution for others to retrieve.

Thank you!
Peng

**Comment by JoshData** [comment]
IMO, DNSSEC and GPG are dying technologies and GPG in particular is extremely hard to do right in a novice-friendly environment like the one that Mail-in-a-Box strives for, so I don't think I would accept PRs for either of these features.

**Comment by ddavness** [comment]
I'd also argue that an issue of DNS records to serve PGP keys is that they can become way too big (beyond what's acceptable, in my opinion, for a record).

On the other hand, WKD is still [technically a draft](https://tools.ietf.org/id/draft-koch-openpgp-webkey-service-11.html). While it hasn't changed much across the last 1.5 years, one can't quite consider it a stable thing yet.

This said, (_shameless plug here, I am sorry_) I am actually hacking around with some form of WKD distribution on my [fork](https://github.com/ddavness/power-mailinabox/tree/dev-wkd). Still far from a functional product, though (I still need to answer myself some design questions around it).

**Comment by iuvbio** [comment]
I think an easy way to do this yourself would be to use keys.openpgp.org 's [WKD as a service](https://keys.openpgp.org/about/usage#wkd-as-a-service). I haven't tested it myself yet, but it seems like setting a custom CNAME DNS record to  wkd.keys.openpgp.org should be enough.

**Comment by githubatf2f10** [comment]
Thanks all for the input! Thanks Josh for your thoughts!
I just put some notes about what I did on my box for future reference for myself or some others.
WKD:
Advanced method: (I don't know how to do this in MAIB).
This needs a subdomain, like openpgpkey.abc.com...
This needs to have a structure like,
            .well-known/openpgpkey/abc.com/....

Direct method: (I manually did this on my own abc.com's root directory)
/home/user-data/www/abc.com$
           .well-known/openpgpkey/
            hu/all the keys here.....  
            policy
sudo chown -R user-data:root .well-known/

Add following to nginx-alldomains.conf:
	location ^~ /.well-known/openpgpkey {
		default_type "application/octet-stream";
		add_header Access-Control-Allow-Origin * always;
	}
do web-update:

**Comment by ValdikSS** [comment]
@JoshData, I agree that PGP in email is/was not straightforward to use and considered an advanced feature that is used by privacy-oriented tech-savvy users, but there were some upgrades of PGP key discovery infrastructure in the last years. For example, now you don't need to send the key or give an URL to the recipient beforehand, it could be discovered automatically via OpenPGP Web Key Directory (WKD) — just publish your key on the webserver of your domain in a special format and the email client would request it.

There's also newer key server infrastructure available, https://keys.openpgp.org/ for example. Add your key there and many email clients/providers with PGP support would discover and use it automatically.

Newer versions of Thunderbird now have built-in PGP support, works pretty good. I receive automatically encrypted emails from Protonmail users occasionally, and can reply encrypted as well.

**Comment by githubatf2f10** [comment]
https://wiki.gnupg.org/WKDHosting
https://wiki.gnupg.org/WKD

Since MAIB can support multiple domains and I would think Advanced Method for OpenPGP WKD works out best. We may only have created a default structure, like this, (This can be aliased in Nginx conf, probably "nginx-alldomains.conf")
/home/user-data/openpgpkey                ---->aliased to "/ .well-known/openpgpkey" for all domains.
in /home/user-data, similar to ssl folder where we host each domain's ssl certs.

Whenever user adds a domain, there will autopopulate the followings,
/home/user-data/openpgpkey/abc.com/hu/
/home/user-data/openpgpkey/abc.com/policy
...
/home/user-data/openpgpkey/def.com/hu/
/home/user-data/openpgpkey/def.com/policy

Within each of these directories, are public gpg keys for those domains. (user can manually SCP into or we can add this function into Admin portal, if possible to deal with Read/Write attributes automatically)
https://openpgpkey.abc.com/.well-known/openpgpkey/abc.com/hu/
...
https://openpgpkey.abc.com/.well-known/openpgpkey/abc.com/hu/

**Comment by ddavness** [comment]
With the infrastructure MIAB provides right now, you can easily set up a static WKD via direct method, provided you are willing to manually update the keys:

Edit the `/home/user-data/www/example.com.conf` file and add in the following:
```nginx
location = /.well-known/openpgpkey/policy {
        return 204; # 204 means No Content, so a 0 byte file
}

location ~ ^/.well-known/openpgpkey/hu/([ybndrfg8ejkmcpqxot1uwisza345h769]*)$ {
        alias /var/lib/mailinabox/wkd/davness.dev/$1;
        add_header Access-Control-Allow-Origin "*" always;
}
```
